File: flink-table-1.7.1/src/main/java/WordCountTable.java
import org.apache.flink.api.java.DataSet;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.TableEnvironment;
import org.apache.flink.table.api.java.BatchTableEnvironment;
import util.WC;

public class WordCountTable {
    public static void main(String[] args) throws Exception {
        //问题：CollectionsEnvironment是什么环境？
        //Environment
        ExecutionEnvironment env = ExecutionEnvironment.createCollectionsEnvironment();
        BatchTableEnvironment tableEnv = TableEnvironment.getTableEnvironment(env);
        //Source
        DataSet<WC> input = env.fromElements(
                new WC("Hello", 1),
                new WC("Ciao", 1),
                new WC("Hello", 1));
        //Table
        Table table = tableEnv.fromDataSet(input);
        //SQL
        Table filtered = table.groupBy("word").select("word, frequency.sum as frequency").orderBy("word");//.filter("frequency = 2");
        //DataSet
        DataSet<WC> result = tableEnv.toDataSet(filtered, WC.class);
        //Sink
        result.print();
    }
}

File: flink-table-1.7.1/src/main/scala/scala/WordCountTable.scala
package scala

import org.apache.flink.api.scala.ExecutionEnvironment
import org.apache.flink.api.scala._
import org.apache.flink.table.api.TableEnvironment
import org.apache.flink.table.api.scala._

import scala.util.WC

object WordCountTable {
  def main(args: Array[String]): Unit = {
    //Environment
    val env = ExecutionEnvironment.getExecutionEnvironment
    val tableEnv = TableEnvironment.getTableEnvironment(env)
    //Source
    val input = env.fromElements(WC("hello", 1), WC("hello", 1), WC("ciao", 1))
    //Table
    val table = input.toTable(tableEnv)
    //SQL
    //列名表示方法：'NAME
    val filtered = table.groupBy('word).select('word, 'frequency.sum as 'frequency).filter('frequency === 2)
    //DataSet
    val result = filtered.toDataSet[WC]
    //Sink
    result.print()
  }
}
